dataline-workers/src/main/java/io/dataline/workers/protocol/singer/SingerJsonStreamFactory.java

```diff
@@ -24,58 +24,50 @@
 
 package io.dataline.workers.protocol.singer;
 
-import com.fasterxml.jackson.core.JsonParser;
-import com.fasterxml.jackson.core.JsonToken;
 import com.fasterxml.jackson.databind.ObjectMapper;
-import com.google.common.collect.AbstractIterator;
 import io.dataline.config.SingerMessage;
+import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.Objects;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-public class SingerJsonIterator extends AbstractIterator<SingerMessage> {
+/**
+ * Creates a stream from an input stream. The produced stream attempts to parse each line of the
+ * InputStream into a SingerMessage. If the line cannot be parsed into a SingerMessage it is
+ * dropped. Each record MUST be new line separated.
+ *
+ * <p>If a line starts with a SingerMessage and then has other characters after it, that
+ * SingerMessage will still be parsed. If there are multiple SingerMessage records on the same line,
+ * only the first will be parsed.
+ */
+public class SingerJsonStreamFactory {
```

Comment:
This is not a factory since the inputstream is consumed. What is the rational for removing the iterator? Everything about this class says that it should be an iterator.

Reply:
If you need to filter then add that in the compute next method

Reply:
here are the two versions of this that i tried as iterators. neither of them read better to me:

1. 
```
public class SingerJsonIterator implements Iterator<SingerMessage> {
  private static final Logger LOGGER = LoggerFactory.getLogger(SingerJsonIterator.class);
  private final ObjectMapper objectMapper;
  private final Iterator<SingerMessage> iterator;

  public SingerJsonIterator(InputStream inputStream) {
    final BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
    this.objectMapper = new ObjectMapper();
    iterator =
        bufferedReader.lines().map(this::parseJsonOrNull).filter(Objects::nonNull).iterator();
  }

  private SingerMessage parseJsonOrNull(String record) {
    try {
      LOGGER.info("record = " + record);
      return objectMapper.readValue(record, SingerMessage.class);
    } catch (IOException e) {
      LOGGER.info(
          String.format(
              "Record was not a json representation of a SingerMessage. Record: %s", record),
          e);
      return null;
    }
  }

  @Override
  public boolean hasNext() {
    return iterator.hasNext();
  }

  @Override
  public SingerMessage next() {
    return iterator.next();
  }
}
```

2. 
```
public class SingerJsonIterator2 extends AbstractIterator<SingerMessage> {
  private static final Logger LOGGER = LoggerFactory.getLogger(SingerJsonIterator2.class);
  private final ObjectMapper objectMapper;
  private final Iterator<String> internalIterator;

  public SingerJsonIterator2(InputStream inputStream) {
    this.internalIterator =
        new BufferedReader(new InputStreamReader(inputStream)).lines().iterator();
    this.objectMapper = new ObjectMapper();
  }

  private SingerMessage parseJsonOrNull(String record) {
    try {
      LOGGER.info("record = " + record);
      return objectMapper.readValue(record, SingerMessage.class);
    } catch (IOException e) {
      LOGGER.info(
          String.format(
              "Record was not a json representation of a SingerMessage. Record: %s", record),
          e);
      return null;
    }
  }

  @Override
  protected SingerMessage computeNext() {
    while (internalIterator.hasNext()) {
      final SingerMessage singerMessageOptional = parseJsonOrNull(internalIterator.next());
      if (singerMessageOptional != null) {
        return singerMessageOptional;
      }
    }
    return endOfData();
  }
}
```

I take it you'd prefer 2?

Reply:
actually why not a mix of 1 and 2? just define the internal iterator in 2 the same way you define it in one so you're compute Next becomes trivial

Reply:
TBH they don't have to read well. Iterator are here to hide complexity. As long as the consumer looks clean

Reply:
> TBH they don't have to read well. Iterator are here to hide complexity. As long as the consumer looks clean

this argument doesn't make sense to me in this context. we are deciding between exposing a stream or an iterator. both are effectively the same to the consumer. both are hiding the complexity equally. in fact, the iterator is trivially less good since, in practice we are immediately going to convert it into a stream ([here](https://github.com/datalineio/dataline/blob/master/dataline-workers/src/main/java/io/dataline/workers/singer/SingerTapFactory.java#L125)). pretty much the only thing that matters between the two options we are considering is producing easy to read source.

Reply:
put up the iterator version.